Mathlib/Order/Minimal.lean

```diff
@@ -1,507 +1,645 @@
+  ⟨⟨hQ, h.prop⟩, fun _ hy ↦ h.le_of_le hy.2⟩
+
+theorem Maximal.and_right (h : Maximal P x) (hQ : Q x) : Maximal (fun x ↦ (P x ∧ Q x)) x :=
+  ⟨⟨h.prop, hQ⟩, fun _ hy ↦ h.le_of_ge hy.1⟩
+
+theorem Maximal.and_left (h : Maximal P x) (hQ : Q x) : Maximal (fun x ↦ (Q x ∧ P x)) x :=
+  ⟨⟨hQ, h.prop⟩, fun _ hy ↦ h.le_of_ge hy.2⟩
+
+@[simp] theorem minimal_eq_iff : Minimal (· = y) x ↔ x = y := by
+  simp (config := {contextual := true}) [Minimal]
+
+@[simp] theorem maximal_eq_iff : Maximal (· = y) x ↔ x = y := by
+  simp (config := {contextual := true}) [Maximal]
+
+theorem Minimal.or (h : Minimal (fun x ↦ P x ∨ Q x) x) : Minimal P x ∨ Minimal Q x := by
+  obtain ⟨(h | h), hmin⟩ := h
+  · exact .inl ⟨h, fun y hy hyx ↦ hmin (Or.inl hy) hyx⟩
+  exact .inr ⟨h, fun y hy hyx ↦ hmin (Or.inr hy) hyx⟩
+
+theorem Maximal.or (h : Maximal (fun x ↦ P x ∨ Q x) x) : Maximal P x ∨ Maximal Q x :=
+  Minimal.or (α := αᵒᵈ) h
+
+theorem minimal_and_iff_right_of_imp (hPQ : ∀ ⦃x⦄, P x → Q x) :
+    Minimal (fun x ↦ P x ∧ Q x) x ↔ (Minimal P x) ∧ Q x := by
+  simp_rw [and_iff_left_of_imp (fun x ↦ hPQ x), iff_self_and]
+  exact fun h ↦ hPQ h.prop
+
+theorem minimal_and_iff_left_of_imp (hPQ : ∀ ⦃x⦄, P x → Q x) :
+    Minimal (fun x ↦ Q x ∧ P x) x ↔ Q x ∧ (Minimal P x) := by
+  simp_rw [iff_comm, and_comm, minimal_and_iff_right_of_imp hPQ, and_comm]
+
+theorem maximal_and_iff_right_of_imp (hPQ : ∀ ⦃x⦄, P x → Q x) :
+    Maximal (fun x ↦ P x ∧ Q x) x ↔ (Maximal P x) ∧ Q x :=
+  minimal_and_iff_right_of_imp (α := αᵒᵈ) hPQ
+
+theorem maximal_and_iff_left_of_imp (hPQ : ∀ ⦃x⦄, P x → Q x) :
+    Maximal (fun x ↦ Q x ∧ P x) x ↔ Q x ∧ (Maximal P x) :=
+  minimal_and_iff_left_of_imp (α := αᵒᵈ) hPQ
+
+end LE
+
+section Preorder
+
+variable [Preorder α]
+
+theorem minimal_iff_forall_lt : Minimal P x ↔ P x ∧ ∀ ⦃y⦄, y < x → ¬ P y := by
+  simp [Minimal, lt_iff_le_not_le, not_imp_not, imp.swap]
+
+theorem maximal_iff_forall_lt : Maximal P x ↔ P x ∧ ∀ ⦃y⦄, x < y → ¬ P y :=
+  minimal_iff_forall_lt (α := αᵒᵈ)
+
+theorem Minimal.not_prop_of_lt (h : Minimal P x) (hlt : y < x) : ¬ P y :=
+  (minimal_iff_forall_lt.1 h).2 hlt
+
+theorem Maximal.not_prop_of_lt (h : Maximal P x) (hlt : x < y) : ¬ P y :=
+  (maximal_iff_forall_lt.1 h).2 hlt
+
+theorem Minimal.not_lt (h : Minimal P x) (hy : P y) : ¬ (y < x) :=
+  fun hlt ↦ h.not_prop_of_lt hlt hy
+
+theorem Maximal.not_lt (h : Maximal P x) (hy : P y) : ¬ (x < y) :=
+  fun hlt ↦ h.not_prop_of_lt hlt hy
+
+@[simp] theorem minimal_le_iff : Minimal (· ≤ y) x ↔ x ≤ y ∧ IsMin x :=
+  minimal_iff_isMin (fun _ _ h h' ↦ h'.trans h)
+
+@[simp] theorem maximal_ge_iff : Maximal (y ≤ ·) x ↔ y ≤ x ∧ IsMax x :=
+  minimal_le_iff (α := αᵒᵈ)
+
+@[simp] theorem minimal_lt_iff : Minimal (· < y) x ↔ x < y ∧ IsMin x :=
+  minimal_iff_isMin (fun _ _ h h' ↦ h'.trans_lt h)
+
+@[simp] theorem maximal_gt_iff : Maximal (y < ·) x ↔ y < x ∧ IsMax x :=
+  minimal_lt_iff (α := αᵒᵈ)
+
+theorem not_minimal_iff (hx : P x) : ¬ Minimal P x ↔ ∃ y, y < x ∧ P y := by
```

Comment:
Can you make `not_minimal_iff` be the (currently missing) preorder version, and `not_minimal_iff_exists_lt` be this lemma?